events/events.go

```diff
@@ -0,0 +1,126 @@
+package events
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/hashicorp/serf/serf"
+	"github.com/livepeer/catalyst-api/balancer/catalyst"
+	"github.com/livepeer/catalyst-api/cluster"
+	"github.com/livepeer/catalyst-api/log"
+	"time"
+)
+
+const streamEventResource = "stream"
+const nukeEventResource = "nuke"
+const nodeStatsEventResource = "nodeStats"
+const nodeStreamsEventResource = "nodeStreams"
+
+type Event interface{}
+
+type GenericEvent struct {
+	Resource string `json:"resource"`
+}
+
+type StreamEvent struct {
+	Resource   string `json:"resource"`
+	PlaybackID string `json:"playback_id"`
+}
+
+type NukeEvent struct {
+	Resource   string `json:"resource"`
+	PlaybackID string `json:"playback_id"`
+}
+
+type NodeStatsEvent struct {
+	Resource      string               `json:"resource"`
+	NodeID        string               `json:"node_id"`
+	NodeMetrics   catalyst.NodeMetrics `json:"node_metrics"`
+	NodeLatitude  float64              `json:"node_latitude"`
+	NodeLongitude float64              `json:"node_longitude"`
+}
+
+type NodeStreamsEvent struct {
+	Resource string                     `json:"resource"`
+	NodeID   string                     `json:"node_id"`
+	Streams  map[string]catalyst.Stream `json:"streams"`
+}
+
+func Unmarshal(payload []byte) (Event, error) {
+	var generic GenericEvent
+	err := json.Unmarshal(payload, &generic)
+	if err != nil {
+		return nil, err
+	}
+	switch generic.Resource {
+	case streamEventResource:
+		event := &StreamEvent{}
+		err := json.Unmarshal(payload, event)
+		if err != nil {
+			return nil, err
+		}
+		return event, nil
+	case nukeEventResource:
+		event := &NukeEvent{}
+		err := json.Unmarshal(payload, event)
+		if err != nil {
+			return nil, err
+		}
+		return event, nil
+	case nodeStatsEventResource:
+		event := &NodeStatsEvent{}
+		err := json.Unmarshal(payload, event)
+		if err != nil {
+			return nil, err
+		}
+		return event, nil
+	case nodeStreamsEventResource:
+		event := &NodeStreamsEvent{}
+		err := json.Unmarshal(payload, event)
+		if err != nil {
+			return nil, err
+		}
+		return event, nil
+	}
+	return nil, fmt.Errorf("unable to unmarshal event, unknown resource '%s'", generic.Resource)
+}
+
+func StartMetricSending(nodeName string, latitude float64, longitude float64, c cluster.Cluster) {
+	ticker := time.NewTicker(5 * time.Second)
+	go func() {
+		for range ticker.C {
+			log.LogNoRequestID("catabalancer sending node stats")
+			sysinfo, err := catalyst.GetSystemInfo()
+			if err != nil {
+				log.LogNoRequestID("catabalancer failed to get sys info", "err", err)
+				break
+			}
+			log.LogNoRequestID("catabalancer sys info ", "cpus", len(sysinfo.CPUInfo))
+
+			event := NodeStatsEvent{
+				Resource: "nodeStats",
+				NodeID:   nodeName,
+				NodeMetrics: catalyst.NodeMetrics{
+					CPUUsagePercentage:       int64(sysinfo.LoadInfo.Load5 * 100), // TODO how can we convert load into a percentage? i think we need to take into account the number of cpus
```

Comment:
Load and CPU percentage usage are different metrics.
[Load and https://superuser.com/questions/1402079/linux-understanding-load-average-and-cpu](https://superuser.com/questions/1402079/linux-understanding-load-average-and-cpu)

You're probably looking for https://pkg.go.dev/github.com/shirou/gopsutil/v3@v3.23.10/cpu#Percent

Reply:
Yeah we need both CPU usage (to determine current cpu load) and load averages (to determine if load is going up or down). @mjh1 I can fix this up tomorrow. Will also add something for the bandwidth usage bits.